### PR TITLE
Optimize causal Conv3d VAE padding

### PR DESCRIPTION
**Pull request by mickqian**
## Motivation

Video/image VAEs with causal Conv3d feature caches currently build the conv input with a separate temporal `torch.cat` and spatial `F.pad` when a cached temporal context is present. This shows up in decode-heavy Wan paths as small but repeatable overhead, and the same input-preparation pattern is also used by Qwen Image VAE.

This PR keeps the optimization narrow: it only fuses cached temporal concat plus constant spatial padding for guarded contiguous 5D NCDHW tensors. It does not change parallel decode policy, performance baselines, tiled decode behavior, or the previous halo-buffer experiment.

Wan models are offline generation models, but their VAE decode still uses feature-cache chunking by default for multi-frame video. In that path, each latent temporal slice is decoded separately and later slices consume cached causal-conv boundary features from earlier slices, so this fused path is exercised by normal Wan video decode.

## Changes

- Add a Triton helper that fuses cached causal-conv concat and spatial padding for contiguous 5D NCDHW tensors.
- Add `causal_conv3d_cat_pad(...)` as the shared guarded helper in `parallel_conv.py`, with fallback to the existing PyTorch path.
- Use the helper in:
  - `WanCausalConv3d`
  - `SpatialParallelCausalConv3d`
  - `QwenImageCausalConv3d`

## Validation

Remote H200 validation, based on `origin/main` `ec36dde58083aca8f26c3740332498a11a06debf`.

Microbench, bf16, 100 iterations, comparing old `torch.cat + F.pad` vs fused helper:

| Shape `[B,C,T,H,W]` | baseline ms | fused ms | speedup | max diff |
|---|---:|---:|---:|---:|
| `[1,128,1,60,104]` | 0.0263 | 0.0256 | 1.03x | 0 |
| `[1,256,1,60,104]` | 0.0324 | 0.0229 | 1.41x | 0 |
| `[1,512,1,30,52]` | 0.0265 | 0.0233 | 1.14x | 0 |
| `[1,512,3,30,52]` | 0.0293 | 0.0226 | 1.29x | 0 |
| `[1,256,1,90,160]` | 0.0719 | 0.0305 | 2.35x | 0 |

Real-server `wan2_1_t2v_1.3b_cfg_parallel`, 2x H200:

| Run | Variant | E2E ms | Avg denoise ms | Decode ms | Consistency |
|---|---|---:|---:|---:|---|
| 1 | base | 4805.16 | 79.58 | 483.96 | pass |
| 1 | patched | 4528.75 | 74.39 | 466.25 | pass |
| 2 | base | 4536.84 | 74.21 | 484.35 | pass |
| 2 | patched | 4522.41 | 74.29 | 465.12 | pass |

Final patched-only smoke for the same case after code cleanup:

| Variant | E2E ms | Avg denoise ms | Decode ms | Consistency |
|---|---:|---:|---:|---|
| final_patch | 4523.84 | 74.31 | 465.30 | pass |

Additional `fastwan2_2_ti2v_5b`, 1x H200 spot check:

| Variant | E2E ms | Avg denoise ms | Decode ms | Consistency |
|---|---:|---:|---:|---|
| base | 2504.90 | 84.27 | 1375.55 | pass |
| patched | 2508.88 | 85.82 | 1365.84 | pass |

After review fix `15dd1fe313`, the kernel uses runtime scalar shape/padding arguments plus integer-safe coordinate clamping. H200 microbench, bf16, 100 iterations:

| Shape `[B,C,T,H,W]` | baseline ms | fused ms | speedup | max diff |
|---|---:|---:|---:|---:|
| `[1,128,1,60,104]` | 0.0286 | 0.0230 | 1.246x | 0 |
| `[1,256,1,60,104]` | 0.0299 | 0.0240 | 1.242x | 0 |
| `[1,512,1,30,52]` | 0.0270 | 0.0224 | 1.205x | 0 |
| `[1,512,3,30,52]` | 0.0267 | 0.0292 | 0.916x | 0 |
| `[1,256,1,90,160]` | 0.0708 | 0.0518 | 1.368x | 0 |

The Wan/Qwen feature-cache decode path exercises the `T=1` chunked shape, where the fused path remains faster after the review fix.

Qwen Image patched-only smoke, 2x H200:

| Case | E2E ms | Avg denoise ms | Decode ms | Consistency |
|---|---:|---:|---:|---|
| `qwen_image_t2i_2_gpus` | 11525.21 | 224.46 | 40.24 | pass |

Local pre-commit:

```text
pre-commit run --files python/sglang/jit_kernel/diffusion/triton/causal_conv3d_pad.py python/sglang/multimodal_gen/runtime/layers/parallel_conv.py python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py
```

Passed.

---
### CI States

Latest PR Test (Base): :x: [Run #27526608776](https://github.com/sgl-project/sglang/actions/runs/27526608776)
Latest PR Test (Extra): :x: [Run #27526608699](https://github.com/sgl-project/sglang/actions/runs/27526608699)

### CODE REVIEW

File: python/sglang/jit_kernel/diffusion/triton/causal_conv3d_pad.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    x_t = src_t - cache_t
+    x_offsets = (((ob * channels + oc) * t_size + x_t) * h_size + ih) * w_size + iw
+    cache_offsets = (
+        ((ob * channels + oc) * cache_t + src_t) * h_size + ih
+    ) * w_size + iw
+
+    x_vals = tl.load(x_ptr + x_offsets, mask=valid & ~from_cache, other=0.0)
+    cache_vals = tl.load(cache_ptr + cache_offsets, mask=valid & from_cache, other=0.0)
+    vals = tl.where(from_cache, cache_vals, x_vals)
+    tl.store(out_ptr + offsets, vals, mask=mask)
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/jit_kernel/diffusion/triton/causal_conv3d_pad.py:70]: Addressed in `15dd1fe313`: dynamic shape/padding values are now runtime scalar arguments, and the load coordinates are clamped before pointer offset computation. I used integer `tl.minimum(tl.maximum(...))` instead of `tl.clamp` because the Triton version in our H200 env rejects integer `tl.clamp`. Re-ran the fused-vs-baseline microbench on H200 with `max_diff=0`.

File: python/sglang/multimodal_gen/runtime/layers/parallel_conv.py [hunk 2]
```diff
@@ -14,6 +14,13 @@
 )
 from sglang.multimodal_gen.runtime.platforms import current_platform
 
+if current_platform.is_cuda():
+    from sglang.jit_kernel.diffusion.triton.wan_vae_pad import (
+        fused_causal_conv3d_cat_pad,
+    )
+else:
+    fused_causal_conv3d_cat_pad = None
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/layers/parallel_conv.py:22]: Keeping the import strict here. CUDA diffusion execution depends on Triton; silently swallowing `ImportError` would hide an invalid install and disable this path unexpectedly. The helper already falls back at runtime for unsupported tensor/layout conditions, but a missing Triton package should remain an environment error.
